### PR TITLE
fix(3258): handle no prototype objects in R.clone

### PR DESCRIPTION
**Pull request by thefourtheye**
Fixes: https://github.com/ramda/ramda/issues/3258

----
cc: @CrossEye @customcommander @denishowew @semmel

### CODE REVIEW

File: source/internal/_clone.js [hunk 1]
```diff
@@ -30,7 +30,8 @@ export default function _clone(value, deep, map) {
     map.set(value, copiedValue);
 
     for (var key in value) {
-      if (value.hasOwnProperty(key)) {
+      // If there is no prototype, the property must be its own
+      if (!_hasPrototype(value) || value.hasOwnProperty(key)) {
```
Comment (CrossEye) [inline source/internal/_clone.js:34]: This is not precisely an objection.  I'm curious why you chose to implement it this way rather than as in this (untested) snippet:

```js
var hasOwnProperty = Object .prototype .hasOwnProperty;
/* ... */

for (var key in value) {
    if (hasOwnProperty .call (value, key) {
/* ... */
```

File: source/internal/_clone.js [hunk 2]
```diff
@@ -83,7 +88,11 @@ _ObjectMap.prototype.set = function(key, value) {
 _ObjectMap.prototype.hash = function(key) {
   let hashedKey = [];
   for (var value in key) {
-    hashedKey.push(key[value]);
+    if (_hasPrototype(key[value])) {
+      hashedKey.push(key[value]);
+    } else {
+      hashedKey.push(Object.prototype.toString.call(key[value]));
+    }
```
Comment (CrossEye) [inline source/internal/_clone.js:95]: What is the use-case that requires this block?  I'm not following something.

**Review by CrossEye — COMMENTED** [review]
I raised two questions inline.

File: source/internal/_clone.js (hunk 2 above)
Comment (thefourtheye) [reply source/internal/_clone.js:95]: @CrossEye All the values are added in `hashedKey` and `hashedKey.join` tries to convert all the elements by calling `.toString` on it, if necessary. In our case, the object without prototype doesn't have `toString`. So, handling that case with `Object.prototype.toString.call`. I was not sure about the implications of doing `Object.prototype.toString.call` on all the elements of the `hashedKey`. That's why it is done this way.

File: test/clone.js [hunk 3]
```diff
@@ -299,6 +299,23 @@ describe('deep clone edge cases', function() {
 
 });
 
+describe('clone objects with no prototypes', function() {
+  it('object with no prototype', function() {
```
[the remark this answers is not in the document]
Comment (thefourtheye) [reply test/clone.js:303]: @customcommander You are correct. The second one is like a super set. Let me drop this.

File: source/internal/_clone.js (hunk 1 above)
Comment (thefourtheye) [reply source/internal/_clone.js:34]: @CrossEye When this patch was created, I was a bit concerned about the readability. Hence the comment as well. I am all up for doing this the concise way. Let me update this.

**Review by CrossEye — APPROVED** [review]
LGTM

:herb:

**Review by customcommander — APPROVED** [review]

**Comment by David-Hobson** [comment]
Came across this issue recently and am glad to see that it's fixed. Do we know when this going to get released and be a part of the latest Ramda package on npm?
